### PR TITLE
extract `model.pushObjects` to a method

### PR DESCRIPTION
**Pull request by asross**
...to allow overrides in the case when more specialized munging/manipulation might be required.

For one part of our app that needs infinite scroll, we have a slightly non-traditional API that requires a more complicated merging of our initial data and subsequent requests. This extremely minor change makes it much easier for us to use `ember-infinity` in that case.

### CODE REVIEW

File: addon/mixins/route.js [hunk 1]
```diff
@@ -217,6 +217,10 @@ export default Ember.Mixin.create({
     return false;
   },
 
+  _updateModel(model, infinityModel) {
```
Comment (kellyselden) [inline addon/mixins/route.js]: If the purpose is overriding, wouldn't you want to make it public?

Comment (hhff) [inline addon/mixins/route.js]: Yeah agree!  I was under the impression that the "pushObjects" to model content no longer works in new versions of Ember Data.  What version of Ember Data are you using?

Comment (asross) [inline addon/mixins/route.js]: Will un-underscore the method name. I'm not sure about the compatibility of `pushObjects` with the latest Ember data, but that doesn't seem relevant to this change (since I'm just moving code around).

**Comment by kellyselden** [comment]
I would suggest a README change to help others with this. Then squash commits and LGTM.

File: addon/mixins/route.js [hunk 2]
```diff
@@ -217,6 +217,10 @@ export default Ember.Mixin.create({
     return false;
   },
 
+  updateModel(model, infinityModel) {
```
Comment (hhff) [inline addon/mixins/route.js]: We also don't need to pass `infinityModel` here, we can get it with `this.get('_modelPath')`.

Can we also rename this method to `addObjects`, and the first arg to `objects`?  Want to be explicit that we aren't updating or changing the infinityModel in place, we're just adding to it.

Also - a small note in the README would be great!

Comment (kellyselden) [inline addon/mixins/route.js]: @hhff I think the point of the `infinityModel` param is to use it when overriding without accessing anything private.

Comment (hhff) [inline addon/mixins/route.js]: Its fine to use private methods inside a public method?

Comment (kellyselden) [inline addon/mixins/route.js]: Are you saying that when overriding, people should use `_modelPath` to get the `infinityModel`?

Comment (hhff) [inline addon/mixins/route.js]: no, im saying we don't even need the 2nd arg, cause we can retrieve it from the route

Comment (kellyselden) [inline addon/mixins/route.js]: Hmm. I think you know better than me in this area, I'm going to bow out.

Comment (hhff) [inline addon/mixins/route.js]: I may be missing something!  The purpose of this method is just for conveniently pushing arbitary objects into the infinityModel from the route, right @asross ?

Comment (asross) [inline addon/mixins/route.js]: Sorry for being so late to reply! I'd be in favor of keeping `infinityModel` as a parameter because we don't necessarily want users of this addon to have to think deeply about implementation details of `ember-infinity`, instead we just want them to specify how the response from the server (`infinityModel`) should update the route's data (`model`). So from an addon-overrider user experience point of view, I think it's slightly nicer to have it be an argument.

Another argument-argument is that you might change the private variable name later, but you don't necessarily want users of the addon to have to change their code.

But, from a general software point of view, it does make more sense to omit the second argument since it _is_ superfluous. And if you're overriding this method, even though it's public, you maybe know what you're in for.

In any case, I'm happy with either, the main thing is to have a separate method. I'll add something to the README to this effect and whichever you think is better @hhff  we will go with.

Comment (asross) [inline addon/mixins/route.js]: Actually, I'm wondering if we should rename this method to something more specific to `ember-infinity`? It's conceivable that an app could already have a route method called `updateModel` that's unrelated to this addon.

Comment (asross) [inline addon/mixins/route.js]: I changed it to `updateInfinityModel` to avoid name collision, hope that's alright and sorry for inadvertently hiding this chain of comments.

File: addon/mixins/route.js [hunk 3]
```diff
@@ -217,6 +217,10 @@ export default Ember.Mixin.create({
     return false;
   },
 
+  updateInfinityModel(model, infinityModel) {
+    model.pushObjects(infinityModel.get('content'));
+  },
+
```
Comment (hhff) [inline addon/mixins/route.js]: I'm still finding this confusing.  the `infinityModel` should always refer to the route's model, ie, the model stored at `_modelPath`.

I think we should do

```js
updateInfinityModel(collection, infinityModel) {
    if (!infinityModel) {
      infinityModel = this.get('_modelPath');
    }
    return infinityModel.pushObjects(collection.get('content'));
}
```

This would imply we'd want to change the server response in the promise above from `infinityModel` to `collection`, and then line 191 to: `this.updateInfinityModel(collection, model);`

I think the wording `infinityModel` should canonically always refer to the object at `route.get('_modelPath')` and nothing else.

This also means that developers can also use this function like so:

```js
actions: {
  pushHughIntoInfinityModel() [
    var updatedInfinityModel = this.updateInfinityModel([
      { id: 1, name: "Hugh Francis" }
    ]);
    console.log(updatedInfinityModel);
  }
}
```

...which is something some other devs have asked about.

Comment (asross) [inline addon/mixins/route.js]: :+1:, I really like that. Will make that change and potentially update the README too.

Comment (hhff) [inline addon/mixins/route.js]: Psyched!  sorry for the nitpicks, just think this is an important addition to the API and I wanna get it right xo

**Comment by asross** [comment]
I updated the method signature and also renamed some variables so that `infinityModel` consistently refers to the route's model rather than the new objects to be added to it. Let me know if the README example which I c/ped from you is ok :grin:

File: addon/mixins/route.js [hunk 4]
```diff
@@ -217,6 +216,19 @@ export default Ember.Mixin.create({
...
+   @method updateInfinityModel
+   @param {Ember.Enumerable} newObjects The new objects to add to the model
+   @return {Ember.Array} returns the updated infinity model
+   */
+  updateInfinityModel(newObjects) {
+    var infinityModel = this.get(this.get('_modelPath'));
+
+    return infinityModel.pushObjects(newObjects);
+  },
+
```
Comment (hhff) [inline addon/mixins/route.js]: Ahhh now the problem is that we've pulled the `newObjects.get('content')` back into the load method.  That means, if someone is using pure arrays without prototype extensions, there won't be a good place for them to override the `get('content')` part of the push.

I like this idea because it exposes the method for actually extracting the data from the collection that is pushed into the `infinityModel`, giving developers a public place to change that logic.

Comment (asross) [inline addon/mixins/route.js]: Hmm, but doesn't that mean you would no longer be able to do `updateInfinityModel` with a literal array?

Comment (asross) [inline addon/mixins/route.js]: I think the example you gave wouldn't be possible if we did the `get('content')` inside this method (without an override).

Comment (hhff) [inline addon/mixins/route.js]: It would work for the majority of developers, unless they've explicitly turned off prototype extensions, and if they have, they're probably going to be aware of that stuff.

Comment (asross) [inline addon/mixins/route.js]: Ok, sounds good. Will make the change.

Comment (asross) [inline addon/mixins/route.js]: Done, apologies for the delay.

**Comment by hhff** [comment]
this is perfect @asross - can you put up a simple test?  Once we've got some test coverage this is good to merge

File: README.md [hunk 5]
```diff
@@ -111,6 +111,19 @@ and ember-infinity will be set up to parse the total number of pages from a JSON
 }
 ```
 
+You may override `updateInfinityModel` to customize how the route's `model` should be updated with new objects. You may also invoke this method directly to manually push new objects into the model:
+
+```js
+actions: {
+  pushHughIntoInfinityModel() [
```
Comment (kellyselden) [inline README.md]: nice name choice :)

File: tests/unit/mixins/route-test.js [hunk 6]
```diff
@@ -308,3 +308,70 @@ test('it uses overridden params when loading more data', assert => {
...
+
+  assert.equal(route.get('_canLoadMore'), false);
+  assert.equal(model.get('content.length'), 2);
+  assert.equal(model.get('content.lastObject.author'), 'F. Scott Fitzgerald', 'overrides to updateInfinityModel should take effect');
+
+  var newObjects = Ember.ArrayProxy.create({
+    content: Ember.A([
+      { id: 3, title: 'Tender Is the Night' }
+    ])
+  });
```
Comment (asross) [inline tests/unit/mixins/route-test.js]: This was a little bit laborious to construct... I might be wrong, but I don't think manual invocations of `updateInfinityModel` as specified in your example @hhff will work even with prototype extensions. However, it's still at least possible, and maybe it's nice to have the flexibility of being able to access meta parameters in `updateInfinityModel`.

So my default plan will be to keep this and update the README so that instead of
```js
var updatedInfinityModel = this.updateInfinityModel([
  { id: 1, name: "Hugh Francis" }
]);
```
we have:
```js
var updatedInfinityModel = this.updateInfinityModel(Ember.ArrayProxy.create({
  content: Ember.A([
    { id: 1, name: "Hugh Francis" }
  ])
});
```
Let me know how that sounds or if I just am misunderstanding prototype extensions!

Comment (asross) [inline tests/unit/mixins/route-test.js]: Ok, reading up on it a bit more, I was definitely misunderstanding prototype extensions and some aspects of this addon. Prototype extensions are disabled locally in this project but we expect that most consumers of the addon will have them enabled. So hopefully the example above should work?

The only thing I'm not sure about now is whether an ES6 or ES5 prototype-extended Array will respond properly to `.get('content')`. But I think I don't need to update the README as drastically as I thought. This has been a nice learning experience.

Comment (hhff) [inline tests/unit/mixins/route-test.js]: Yeah!  I think this is ready...!  I think if people want to use non-prototype extension'd collections, we'll just recommend they override the hook to account for that case.

**Comment by hhff** [comment]
Can u give it a quick squash?

**Comment by asross** [comment]
> Can u give it a quick squash?

Yep, done!

**Comment by hhff** [comment]
:+1:

**Comment by kellyselden** [comment]
Nice work.

**Comment by asross** [comment]
Thanks!
